sdk/core/azure-core/inc/azure/core/operation.hpp

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "azure/core/operation_status.hpp"
+#include "azure/core/response.hpp"
+
+#include <chrono>
+
+namespace Azure { namespace Core {
+
+  //@param <T> The type of the final result of the long-running operation.
+  template <class T> class Operation {
+  protected:
+    OperationStatus m_status = OperationStatus::NotStarted;
+  public:
+    OperationStatus Status() { return m_status; }
+
+    bool Done() {
+      return (
+          m_status == OperationStatus::Succeeded ||
+          m_status == OperationStatus::Cancelled ||
+          m_status == OperationStatus::Failed);
+    }
+
+    bool HasValue()
+    {
+      return (m_status == OperationStatus::Succeeded);
+    }
+
+    virtual T Value() = 0;
+
+    virtual std::string GetResumeToken() = 0;
+
+    virtual std::unique_ptr<Http::RawResponse> Poll(Context& context = Context()) = 0;
```

Comment:
I initially was going to use what you described.  Then decided I could be crafty with optional args.  I wasn't aware that wasn't possible with lvalue refs.  Too bad the compiler doesn't warn me about it :).

Reply:
> Too bad the compiler doesn't warn me about it

There's some weirdness here since it's a class template and the function is pure virtual. This:
```c++
struct Context {};

struct S {
  virtual void meow(Context& = Context()) = 0;
};
```
gets a diagnostic (https://godbolt.org/z/fb6vff) although MSVC only complains with `-permissive-` (the "try to be more standard-conforming" flag) since this is notoriously an area where legacy MSVC was broken. (Apparently someone tried to portray this as a feature instead of a bug in ancient times earning it the nickname "Evil Extension".)

Reply:
This is OK to do if the parameter was marked as const, correct, if we want an optional argument with a default value?

We do this quite often for our `Options` types. For example:
https://github.com/Azure/azure-sdk-for-cpp/blob/ed1908efa149e2dd8b567c4a36d4dcc56b754769/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/append_blob_client.hpp#L37-L41

Reply:
> This is OK to do if the parameter was marked as const, correct, if we want an optional argument with a default value?

Yes, a `const` lvalue reference can bind to an rvalue. I've clarified this in the original comment.